New module postgresql_tablespace

Andersson007 opened the pull request:
##### SUMMARY
Add or remove PostgreSQL tablespaces from remote hosts

##### ISSUE TYPE
- New Module Pull Request

##### MAIN OPTIONS
tablespace: Name of the tablespace to add or remove.
location: Path to the tablespace directory in the file system.
state: absent | present
owner: Name of the role to set as an owner of the tablespace.
set: Dict of tablespace options to set. Supported from PostgreSQL 9.0.
rename_to: New name of the tablespace.
session_role: Switch to session_role after connecting.
...

##### RETURN
```
queries:
    description: List of queries that was tried to be executed.
    returned: always
    type: str
    sample: [ "CREATE TABLESPACE bar LOCATION '/incredible/ssd'" ]
tablespace:
    description: Tablespace name.
    returned: always
    type: str
    sample: 'ssd'
owner:
    description: Tablespace owner.
    returned: always
    type: str
    sample: 'Bob'
options:
    description: Tablespace options.
    returned: always
    type: dict
    sample: { 'random_page_cost': 1, 'seq_page_cost': 1 }
location:
    description: Path to the tablespace in the file system.
    returned: always
    type: str
    sample: '/incredible/fast/ssd'
newname:
    description: New tablespace name
    returned: if existent
    type: str
    sample: new_ssd
state:
    description: Tablespace state at the end of execution.
    returned: always
    type: str
    sample: 'present'
```
##### OUTPUT EXAMPLE
tasks:
   - name: postgresql_tablespace
     postgresql_tablespace:
       tablespace: foo
       owner: alice
       state: present
       location: '/var/lib/pgsql/test_tbl0'
       set:
         random_page_cost: 1
         seq_page_cost: 4

```
changed: [spblnx125] => {
    "changed": true, 
    ....
    "location": "/var/lib/pgsql/test_tbl0", 
    "options": {
        "random_page_cost": "1", 
        "seq_page_cost": "4"
    }, 
    "owner": "alice", 
    "queries": [
        "CREATE TABLESPACE foo LOCATION '/var/lib/pgsql/test_tbl0'", 
        "ALTER TABLESPACE foo OWNER TO alice", 
        "ALTER TABLESPACE foo SET (random_page_cost = '1')", 
        "ALTER TABLESPACE foo SET (seq_page_cost = '4')"
    ], 
    "state": "present", 
    "tablespace": "foo"
}
```
##### EXAMPLES
```
- name: Create a new tablespace called acme and set bob as an its owner
  postgresql_tablespace:
    name: acme
    owner: bob
    location: /data/foo

- name: Create a new tablespace called bar with tablespace options
  postgresql_tablespace:
    name: bar
    set:
      random_page_cost: 1
      seq_page_cost: 1

- name: Reset random_page_cost option
  postgresql_tablespace:
    name: bar
    set:
      random_page_cost: reset

- name: Rename the tablespace from bar to pcie_ssd
  postgresql_tablespace:
    name: bar
    rename_to: pcie_ssd

- name: Drop tablespace called bloat
  postgresql_tablespace:
    name: bloat
    state: absent
```

Andersson007 committed:
New module postgresql_tablespace

Andersson007 commented:
+label postgresql
+label database

Andersson007 committed:
New module postgresql_tablespace: sanity fixes

Andersson007 committed:
New module postgresql_tablespace: add tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 committed:
New module postgresql_tablespace: fix tests

Andersson007 commented:
ready_for_review

Andersson007 commented:
@Dorn- @antoinell , Hello, waiting for your painstaking review ;)

Andersson007 commented:
@antoinell , as a co-author of this module, can you find a few time to review it? @Dorn- , you too?
If you can't do it, please, let me know. I'll try to kick somebody else.

antoinell commented:
It is great work with impressive tests. @Andersson007 you did a lot of background research to find out how postgres keeps track of the space location for instance and also about when the space options were introduced.
my comments ::
- `pg_quote` could be used to maintain consistency with the other postgres modules in ansible,
-  in the case of a totally new tablespace creating the tablespace in one command specifying the owner and the options seems more natural than creating the tablespace without specifying the owner, then setting the owner, then setting the options. I prefer to have systems do atomic work whenever possible.
- in line 433 there is a typo `# Change autocommit to Fasle if check_mode:`
- in line 318 `def set_settings(self, new_settings=False)` I wonder why there is a default for the parameter new_settings, on top of that should the default not be an empty dictionary rather than `False` ?
- going back from a tablespace with settings to one without settings or changing which particular setting is set does not seem to be implemented. In the [alter tablespace](https://www.postgresql.org/docs/11/sql-altertablespace.html) documentation a `RESET` keyword is mentioned which should be used to remove settings which are not mentioned any more
- going back from a tablespace with a specific owner to a tablespace with a default owner (is it `postgres` ?) is this implemented currently ?

Andersson007 committed:
New module postgresql_tablespace: fix

Andersson007 commented:
@antoinell , thank you for detailed review.

>in line 433 there is a typo # Change autocommit to Fasle if check_mode:

Fixed

>in line 318 def set_settings(self, new_settings=False) I wonder why there is a default for the parameter new_settings, on top of that should the default not be an empty dictionary rather than False ?

If you pass an empty list instead of False, you catch a sanity test error during CI tests. So, I used False instead of an empty dict too.
However, seems it works. Changed to an empty dict. Hope it'll pass CI

>in the case of a totally new tablespace creating the tablespace in one command specifying the owner and the options seems more natural than creating the tablespace without specifying the owner, then setting the owner, then setting the options. I prefer to have systems do atomic work whenever possible.

Generally, I agree with you and I thought about it too but in this particular case it's more flexible for any cases (create new or just change existing tblspace).
Otherwise, we must to implement several additional methods where they're not actually needed.
So (in this case) I'd insist to separate this work.

>about pg_quote_identifier

I tried to add tablespace=1 to lib/ansible/module_utils/database.py bacause it's a global object like the database and the role values
and then change self.name to pg_quote_identifier(self.name, 'tablespace') in the code but it doesn't work.
If you suggest a working variant, I'll change the code.

>going back from a tablespace with settings to one without settings or changing which particular setting is set does not seem to be implemented. In the alter tablespace documentation a RESET keyword is mentioned which should be used to remove settings which are not mentioned any more

Added (also to CI tests), see the documentation and examples.

---
Seems all point fixed except the suffering function pg_quote_identifier() ;)

Andersson007 committed:
New module postgresql_tablespace: fix

Andersson007 commented:
as I said before CI tests failed because new_settings={}. I removed the default value at all because it's not necessary in this case.

Andersson007 commented:
ready_for_review

Andersson007 committed:
New module postgresql_tablespace: fix

antoinell commented:
Thanks @Andersson007 . Looking at 
https://github.com/ansible/ansible/pull/50568/files I notice that ‘pg_quote identifier‘ was used in creating and dropping the tablespace, strangely not for queries. Can you elaborate on why it did not work to use everywhere ‘pg_quote_identifier’ ? I will try it myself as soon as I can.

Andersson007 committed:
New module postgresql_tablespace: fix

Andersson007 commented:
@antoinell I didn't know that it's used in creating and dropping only.
I just added it and tested (also for altering)
Good job, thank you.

Andersson007 commented:
Seems all your notices are satisfied now (except an atomic job, I described why),
ready for your review ;)

antoinell commented:
Great, +1 from me. :-)

Andersson007 commented:
@antoinell if you agree with the module, please, approve it by writing shipit to a new comment

antoinell commented:
shipit

Andersson007 commented:
@kostiantyn-nemchenko could you please look at it ?

kostiantyn-nemchenko commented:
Yes, I'll give you my feedback soon after taking a closer look at this module today. Want to spend some time testing it.

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+    - Tablespace state.
+    - I(state=present) implies the tablespace must be created if it doesn't exist.
+    - I(state=absent) implies the tablespace must be removed if present.
+      I(state=absent) is mutually exclusive with I(location), I(owner), i(opt).
```
seems like "opt" option is not used in current version of module

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+    - I(state=present) implies the tablespace must be created if it doesn't exist.
+    - I(state=absent) implies the tablespace must be removed if present.
+      I(state=absent) is mutually exclusive with I(location), I(owner), i(opt).
+    - See the Note section for information about check mode restrictions.
```
Note -> Notes

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+      verified to be signed by one of these authorities.
+    type: str
+notes:
+- I(stat=absent) and I(state=present) (the second one if the tablespace doesn't exist) do not
```
state=absent

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+    type: str
+notes:
+- I(stat=absent) and I(state=present) (the second one if the tablespace doesn't exist) do not
+  support check mode because the corresponding PostgreSQl DROP and CREATE TABLESPACE commands
```
PostgreSQl -> PostgreSQL

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+except ImportError:
+    HAS_PSYCOPG2 = False
+
+import ansible.module_utils.postgres as pgutils
```
Unused import?

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+                              "WHERE table_name = 'pg_tablespace' "
+                              "AND column_name = 'spcoptions'", add_to_executed=False)
+
+        # For 9.1 version and earlier:
```
PostgreSQL 9.3 and earlier versions are no longer maintained. Do we really need such conditions in new modules?

Andersson007 replied:
I'm afraid CI tests will fail because they use even 9.0. Anyway it's not difficult to implement this, so I suggest to remain this condition.

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_tablespace.py:
```diff
@@ -0,0 +1,557 @@
+                    changed = self.__reset_setting(i)
+                    self.settings[i] = None
+
+            elif i not in self.settings:
```
Can we rewrite this condition to something like this? `elif (i not in self.settings) or (str(new_settings[i]) != self.settings[i]):`

Andersson007 committed:
New module postgresql_tablespace: fix

Andersson007 commented:
@kostiantyn-nemchenko thank you for the valuable review!
All points fixed (except 9.1 condition, I described why).
ready_for_review

gundalow commented:
Would be useful to click "Resolve conversation" on all done items, that should help us see what's left.
Amaxing tests and review discussion in here, great work everybody!

gundalow commented:
@kostiantyn-nemchenko if you are happy should you please give this a `shipit` comment?

kostiantyn-nemchenko commented:
@gundalow I'm still not happy. Just found few bugs/inconcistencies during manual testing. Will provide @Andersson007 with a detailed feedback in a few minutes.

kostiantyn-nemchenko commented:
@Andersson007 After testing I have some additional information that might be useful for you:
1) [Here](https://github.com/ansible/ansible/blob/ec2b5391df1527011c7d023b784e098875378c85/lib/ansible/modules/database/postgresql/postgresql_tablespace.py#L416) and [here](https://github.com/ansible/ansible/blob/ec2b5391df1527011c7d023b784e098875378c85/lib/ansible/modules/database/postgresql/postgresql_tablespace.py#L47) I would use "set" instead of "options" and "settings". 
Users aren't aware of `settings` variable inside the module's code.

2) You should change error message to "doesn't exist" [here](https://github.com/ansible/ansible/blob/ec2b5391df1527011c7d023b784e098875378c85/lib/ansible/modules/database/postgresql/postgresql_tablespace.py#L474).

3) Module does not honor location option.

Please, consider the following example:
```
- name: Create tablespace test1 with /tmp/test1 location 
  become_user: postgres
  postgresql_tablespace:
    tablespace: test1
    location: /tmp/test1
    state: present
    db: template1

- name: Create tablespace test1 with /tmp/test2 location
  become_user: postgres
  postgresql_tablespace:
    tablespace: test1
    location: /tmp/test2
    state: present
    db: template1
```
The 2nd task just makes sure that `test1` is present though I expected receiving an error/warning that tablespace with name `test1` already exists with different path.

4) Module does not honor default owner value.

Example:
```
- name: Make sure the owner of test1 tablespace is ansible and random_page_cost is set
  become_user: postgres
  postgresql_tablespace:
    tablespace: test1
    location: /tmp/test1
    state: present
    owner: ansible
    set:
      random_page_cost: 2
    db: template1

- name: Reset random_page_cost and return owner back to postgres
  become_user: postgres
  postgresql_tablespace:
    tablespace: test1
    location: /tmp/test1
    state: present
    set:
      random_page_cost: reset
    db: template1
```
The 2nd task does not return tablespace owner back to postgres as stated [here](https://github.com/ansible/ansible/blob/ec2b5391df1527011c7d023b784e098875378c85/lib/ansible/modules/database/postgresql/postgresql_tablespace.py#L55).

Also I have few questions to think about:
    - Should we create a `location` directory if it does not exist? Currently, a task to create tablespace fails in this case.
    - Should we issue a warning message if `rename_to` equals `tablespace`?
    - Should we be consistent about module behaviour in case of missing tablespace? 
      We use `fail_json` when nothing to rename, but `warn` when nothing to drop.

kostiantyn-nemchenko reviewed (changes requested):
Please, take a look at [comment](https://github.com/ansible/ansible/pull/52914#issuecomment-470134643)

Andersson007 commented:
@kostiantyn-nemchenko thank you for detailed review and manual testing!

> - Should we create a location directory if it does not exist? 
IMO, we shouldn't do it, because there is a good module (file) to create the file objects in a suitable place, set permissions, etc. Usually we create dir as root in a mounted volume like /mnt/db/11, set an owner as postgres, permissions to 700, then create a symlink from ~/postgres/11/..., etc. Shortly, better to do it before by file module (more atomic, flexible)

Andersson007 commented:
> 1. Here and here I would use "set" instead of "options" and "settings". Users aren't aware of settings variable inside the module's code.

Good note! It remains from ancient times ;) I'll fix it

Andersson007 committed:
New module postgresql_tablespace: fix by k_n

Andersson007 commented:
> The 2nd task does not return tablespace owner back to postgres as stated here.

1. Yes, it set owner as alice by the first playbook because, regarding the documentation, "If this option is not specified, the tablespace owner is a role that creates the tablespace."
So, Alice created the tablespace and she's an owner.
2. When the second playbook start, tablespace test1 will have already existed.

Andersson007 commented:
> - Should we issue a warning message if rename_to equals tablespace?

IMO, that's enough to see that nothing changed.

Andersson007 commented:
> - Should we be consistent about module behaviour in case of missing tablespace?
Good note, changed it to fail_json

Andersson007 commented:
> 3. Module does not honor location option.
Very good job! Changed, added CI tests

Andersson007 commented:
@kostiantyn-nemchenko , seems fixed for all notes (except that I described why).

ready_for_review

kostiantyn-nemchenko commented:
>>Should we issue a warning message if rename_to equals tablespace?

>IMO, that's enough to see that nothing changed.

Currently, a task with same `tablespace` and `rename_to` will fail to execute alter statement. Outside of that, everything seems to be just fine.

Andersson007 committed:
New module postgresql_tablespace: fix by k_n 2

Andersson007 commented:
Good job! Fixed

kostiantyn-nemchenko commented:
shipit

Andersson007 commented:
@kostiantyn-nemchenko , thank you for the great review! It made this module really better and less problem in the future.

gundalow commented:
@Andersson007 Thank you for the module and tests
@antoinell and @kostiantyn-nemchenko great review and feedback

Merged into `devel` for release in Ansible 2.8

Andersson007 commented:
Thanks all

antoinell commented:
Thanks from me too !
